chain/events/observer.go

```diff
@@ -0,0 +1,255 @@
+package events
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"go.opencensus.io/trace"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+type observer struct {
+	api EventAPI
+
+	gcConfidence abi.ChainEpoch
+
+	ready chan struct{}
+
+	lk        sync.Mutex
+	head      *types.TipSet
+	maxHeight abi.ChainEpoch
+	observers []TipSetObserver
+}
+
+func newObserver(api *cache, gcConfidence abi.ChainEpoch) *observer {
+	obs := &observer{
+		api:          api,
+		gcConfidence: gcConfidence,
+
+		ready:     make(chan struct{}),
+		observers: []TipSetObserver{},
+	}
+	obs.Observe(api.observer())
+	return obs
+}
+
+func (o *observer) start(ctx context.Context) error {
+	go o.listenHeadChanges(ctx)
+
+	// Wait for the first tipset to be seen or bail if shutting down
+	select {
+	case <-o.ready:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (o *observer) listenHeadChanges(ctx context.Context) {
+	for {
+		if err := o.listenHeadChangesOnce(ctx); err != nil {
+			log.Errorf("listen head changes errored: %s", err)
+		} else {
+			log.Warn("listenHeadChanges quit")
+		}
+		select {
+		case <-build.Clock.After(time.Second):
+		case <-ctx.Done():
+			log.Warnf("not restarting listenHeadChanges: context error: %s", ctx.Err())
+			return
+		}
+
+		log.Info("restarting listenHeadChanges")
```

Comment:
this will likely spam a lot, it's once a second with no additional information.

Reply:
Only if we get disconnected because we're reading slowly. Basically, `ChainNotify` buffers internally (16 blocks?), so we'll only see this if we fail to keep up with the chain for 8 minutes.